rest_framework/serializers.py

```diff
@@ -1157,9 +1157,29 @@ def build_relational_field(self, field_name, relation_info):
         field_kwargs = get_relation_kwargs(field_name, relation_info)
 
         to_field = field_kwargs.pop('to_field', None)
-        if to_field and not relation_info.related_model._meta.get_field(to_field).primary_key:
-            field_kwargs['slug_field'] = to_field
-            field_class = self.serializer_related_to_field
+        if to_field:
+            def get_related_field(related_model, to_field):
```

Comment:
Haven't looked into the issue in any depth, but we'd certainly want to factor this out. Eg have it as a function somewhere in `rest_framework.utils`